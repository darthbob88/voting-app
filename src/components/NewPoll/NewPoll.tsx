import React, { useState } from 'react';
import { createPoll } from '../../model/DataAccess';
import { Poll, testPoll } from '../../model/Poll';

function NewPoll() {
    /*  ID: string,
    name: string,
    creator: string,
    creation: number,
    expiration: number,
    method: string,
    candidates: string[],
    */

    const [newPoll, setNewPoll] = useState<Poll>(testPoll);

    const createNewPoll = () => {
        const pollID = newPoll.name.toLocaleLowerCase().replace(/\s/, '-');

        const creator = "darthbob88@gmail.com";
        // TODO: Get some better way to do this. Maybe integrate with Firebase, to set it as part of the data access createPoll function?

        let creation = new Date().getTime();
        var result = new Date();
        result.setDate(result.getDate() + 1);

        let mangledPoll = { ...newPoll, ID: pollID, creation, expiration: result.getTime(), creator };


        createPoll(mangledPoll);
    }
    return (
        <form>
            <label><input type="text"
                onChange={(evt) =>
                    setNewPoll({ ...newPoll, name: evt.target.value })
                }
                value={newPoll.name} /> Please enter a name for your poll</label>


            { // TODO: Get this working with a date picker.
            /* <label><input type="date"
                onChange={(evt) =>
                    setNewPoll({ ...newPoll, expiration: evt.target.value })
                }
                value={newPoll.expiration} /> Please enter an expiration date for your poll</label> */}
            <label><input type="text"
                onChange={(evt) =>
                    setNewPoll({ ...newPoll, name: evt.target.value })
                }
                value={newPoll.name} /> Please enter a name for your poll</label>
            <label><input type="text"
                onChange={(evt) =>
                    setNewPoll({ ...newPoll, name: evt.target.value })
                }
                value={newPoll.name} /> Please enter a name for your poll</label>
        </form>
    );
}

export default NewPoll;