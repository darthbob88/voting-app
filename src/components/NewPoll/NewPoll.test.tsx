import { fireEvent, render, screen } from '@testing-library/react';
import NewPoll from './NewPoll';

const testVoteMethods = [{ value: "FPTP", name: "First past the post" }, { value: "IRV", name: "Instant runoff vote" }];

test('renders New Poll component', () => {
    const { container } = render(<NewPoll />);
    expect(container).toMatchSnapshot();
});

test('Can select poll type', () => {
    render(<NewPoll />);

    const methodSelect = screen.getByTestId('methodSelect');
    const methodOptions: HTMLOptionElement[] = screen.getAllByTestId('methodOption');
    const availableMethods = methodOptions.map(option => option.value);
    testVoteMethods.forEach(method => expect(availableMethods).toContain(method.value));

    fireEvent.change(methodSelect, { target: { value: testVoteMethods[0].value } });

    expect(methodOptions[0].selected).toBeTruthy();
    expect(methodOptions[1].selected).toBeFalsy();
})

// Separate tests for creating polls.
testVoteMethods.forEach(method => test(`Should properly create poll using voting method - ${method}`, async () => {

}))