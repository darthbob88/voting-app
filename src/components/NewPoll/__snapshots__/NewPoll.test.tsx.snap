// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders New Poll component 1`] = `
<div>
  <form>
    <h3>
      New Poll Form
    </h3>
    <label>
      <input
        type="text"
        value="pollName"
      />
       Please enter a name for your poll
    </label>
    <label
      class="label"
    >
      <input
        type="text"
        value="apple,orange,banana"
      />
       Please enter a comma-separated list of candidates for your poll
    </label>
    <label>
      <select
        data-testid="methodSelect"
      >
        <option
          data-testid="methodOption"
          value="FPTP"
        >
          First past the post
        </option>
        <option
          data-testid="methodOption"
          value="IRV"
        >
          Instant runoff vote
        </option>
         
      </select>
       Please select a method for your poll
    </label>
    <button>
      Create poll
    </button>
  </form>
</div>
`;
